ESP32-S3: Restructure LCD_CAM DT layout and add parallel LCD support

xingrz opened the pull request:
The ESP32 LCD_CAM peripheral contains two functional modules: an LCD output block and a DVP camera input block, sharing the same clock and pinctrl resources.

In the current tree, only the camera input path is implemented as a video driver.

This PR restructures the LCD_CAM Devicetree layout by introducing a common `&lcd_cam` node and splitting its functionality into `&lcd_cam_dvp` and `&lcd_cam_disp` child nodes. It also adds support for the LCD output module through a new MIPI-DBI driver.

The new MIPI-DBI driver is enabled on the DNESP32S3B board to drive its I8080-interface LCD.

Related HAL update: https://github.com/zephyrproject-rtos/hal_espressif/pull/501

https://github.com/user-attachments/assets/c5c43adb-9a0c-45d6-b47b-e6b6d0c89fc7

Closes #88385

xingrz committed:
drivers: mipi_dbi: Add ESP32 LCD_CAM backend

The LCD_CAM peripheral is a parallel I/O controller integrated in the ESP32
series. It consists of an LCD output module and a DVP camera input module.
The camera side is already supported in Zephyr as a video driver.

This commit adds support for the LCD module by implementing it as a
MIPI-DBI backend. The LCD_CAM supports multiple output formats, and the
initial implementation provides I8080 mode.

Signed-off-by: Chen Xingyu <hi@xingrz.me>

xingrz committed:
boards: alientek: dnesp32s3b: Enable LCD support

This enables LCD support on the DNESP32S3B board using the ESP32 MIPI-DBI
driver in I8080 mode.

Signed-off-by: Chen Xingyu <hi@xingrz.me>

xingrz committed:
boards: espressif: esp32s3_eye: Update for LCD_CAM restructuring

The former `&lcd_cam` node has been split into an `&lcd_cam` controller
and an `&lcd_cam_dvp` subdevice. Update the board DTS accordingly.

Signed-off-by: Chen Xingyu <hi@xingrz.me>

xingrz committed:
boards: seeed: xiao_esp32s3: Update for LCD_CAM restructuring

The former `&lcd_cam` node has been split into an `&lcd_cam` controller
and an `&lcd_cam_dvp` subdevice. Update the board DTS accordingly.

Signed-off-by: Chen Xingyu <hi@xingrz.me>

xingrz committed:
drivers: video: esp32: Rename to lcd-cam-dvp

Rename the ESP32 LCD_CAM DVP driver and its DT binding from
`espressif,esp32-lcd-cam` to `espressif,esp32-lcd-cam-dvp` to clearly
identify it as the DVP (camera) input block of the LCD_CAM peripheral.

Signed-off-by: Chen Xingyu <hi@xingrz.me>

xingrz committed:
doc: migration-guide-4.4: Update ESP32-S3 LCD_CAM restructuring notes

Add migration guidance for the ESP32-S3 LCD_CAM changes.

Signed-off-by: Chen Xingyu <hi@xingrz.me>

# Conflicts:
#	doc/releases/migration-guide-4.4.rst

JarmouniA commented on boards/alientek/dnesp32s3b/Kconfig.defconfig:
```diff
@@ -0,0 +1,28 @@
+config REGULATOR
+	default y
+
+endif # DISPLAY
```
```suggestion
```

sylvioalves commented on drivers/video/video_esp32_dvp.c:
```diff
@@ -396,6 +391,20 @@ static int video_esp32_init(const struct device *dev)
 	struct video_esp32_data *data = dev->data;
 
+	if (!cfg->cam_clk) {
+		printk("No cam_clk specified\n");
```
Same for the `printk` in this file.

xingrz commented on drivers/mipi_dbi/mipi_dbi_esp32.c:
```diff
@@ -0,0 +1,460 @@
+	}
+
+	if (spi_cs_is_gpio(&dbi_cfg->config) && dbi_cfg->config.cs.gpio.port != NULL) {
+		if (!gpio_is_ready_dt(&dbi_cfg->config.cs.gpio)) {
```
Yeah, just noticed that we can't retrieve the CS GPIO from `dbi_cfg` because we don't have a `spi-dev` phandle pointing to a real SPI controller. Thanks for the
insight!

JarmouniA commented on drivers/mfd/mfd_esp32_lcd_cam.c:
> Where do you think this binding should live?

It can stay where it is, devicetree represents hardware (idealy) independently from software.

xingrz committed:
soc: espressif: Introduce common LCD_CAM support

Move the shared pinctrl and clock initialization for the LCD_CAM peripheral
from the ESP32 DVP driver into a new SoC-level implementation.

Update the ESP32-S3 DTS to reflect the new structure by introducing an
`lcd_cam_dvp` child node under the common `lcd_cam` parent node.

Signed-off-by: Chen Xingyu <hi@xingrz.me>

JarmouniA commented on soc/espressif/common/Kconfig.lcd_cam:
```diff
@@ -0,0 +1,7 @@
+config ESP32_LCD_CAM
+	bool
+	default y if DT_HAS_ESPRESSIF_ESP32_LCD_CAM_DVP_ENABLED
+	default y if DT_HAS_ESPRESSIF_ESP32_LCD_CAM_MIPI_DBI_ENABLED
```
```suggestion
	default y if DT_HAS_ESPRESSIF_ESP32_LCD_CAM_MIPI_DBI_ENABLED
	depends on DT_HAS_ESPRESSIF_ESP32_LCD_CAM_ENABLED
```

JarmouniA commented on doc/releases/migration-guide-4.4.rst:
```diff
@@ -78,6 +78,27 @@ STM32
+
+* The former ``espressif,esp32-lcd-cam`` binding has been restructured. The
+  LCD_CAM peripheral is now represented by a common ``&lcd_cam`` node, with
+  its functional blocks split into two separate child bindings:
```
```suggestion
  its functional blocks split into two separate child nodes:
```

JarmouniA commented on doc/releases/migration-guide-4.4.rst:
```diff
@@ -78,6 +78,27 @@ STM32
+    * :dtcompatible:`espressif,esp32-lcd-cam-mipi-dbi` — the LCD output module,
+      labeled as ``&lcd_cam_disp``.
+
+  The original ``&lcd_cam`` node keeps the common pinctrl, clock, and interrupt
```
```suggestion
  The original :dtcompatible:`espressif,esp32-lcd-cam` compatible node keeps the common pinctrl, clock, and interrupt
```

JarmouniA commented on doc/releases/migration-guide-4.4.rst:
```diff
@@ -78,6 +78,27 @@ STM32
+  Devicetree users must migrate camera-related properties from ``&lcd_cam`` to
+  the new ``&lcd_cam_dvp`` subnode, and update references such as
+  ``zephyr,camera`` to point to ``&lcd_cam_dvp``. LCD-related functionality
+  should now refer to ``&lcd_cam_disp``.
```
```suggestion
  Camera-related properties must be moved from ``lcd_cam`` node to
  the new ``lcd_cam_dvp`` child node, and  ``zephyr,camera`` chosen property should
  point to ``lcd_cam_dvp`` instead.
```

xingrz commented:
Rebased to solve a merging conflict.

xingrz commented:
Rebased again to resolve conflicts.

The change in west.yml is already in the tree, so that commit has been dropped.

sylvioalves commented:
@danieldegrasse Would you take a look in this PR or re-assign if needed? Thanks!

xingrz commented:
Rebased to solve a merge conflict.

xingrz commented:
Rebased to include a CI fix from #100625

ngphibang commented:
`drivers: video: esp32: Rename to lcd-cam-dvp`

For the driver renaming, DVP or MIPI CSI mode can be both supported in the same driver (e.g. see ov5640.c) I think, so `lcd-cam` name is already good, I think, if the HW can support only dvp mode, or in the future if you want to extend it to support also CSI mode.

ngphibang commented:
Ah ok, I see the reason: the renaming is to distinguish the camera module inside the lcd-cam peripheral with the whole peripheral (which can contain other things like lcd-cam-mipi-dbi), so it is OK. In this case, we can rename `lcd-cam-cam` or `lcd-cam-dvp`, both are good, just ignore my comment.

sylvioalves commented:
@xingrz please rebase.

xingrz commented:
Rebased to solve merge conflicts.

josuah commented:
> @josuah, would you take a look?

Just to confirm: I started to give it a look and did not find issue with video APIs.
I was under water while trying to wrap 2025 sorry, and thank you @JarmouniA for taking over the review.
Thanks all for the changes.
